### PR TITLE
Elvis/cell prop

### PR DESCRIPTION
**Pull request by emarc0314**
## Overview

Added navigation and cell tap feedback for going from tracked courses to the detailed list of courses to track.

## Changes Made

Added new functions to NetworkManager and an Endpoint to get from a cell identifier to a specific course page by giving each cell a specific courseID number that can be used to generate a detailed view of courses to track.

## Test Coverage

Tested on phone 

## Screenshots 

<details>

  <summary>Cell Highlighting</summary>

IMG_0225  

</details>

### CODE REVIEW

**Comment by emarc0314** [comment]
gobble gobble if ur reading this happy thanksgiving B)

File: CourseGrab/ViewControllers/Home/HomeTableViewController.swift [hunk 1]
```diff
@@ -295,6 +297,34 @@ extension HomeTableViewController {
...
+    override func tableView(_ tableView: UITableView, didUnhighlightRowAt indexPath: IndexPath) {
+            let cell = tableView.cellForRow(at: indexPath) as! HomeTableViewCell
+            cell.changeBackground(isHighlighted: false)
+    }
+    
+    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
+        switch tableSections[indexPath.section] {
+        case .available(let sections), .awaiting(let sections):
+            NetworkManager.shared.getCourse(courseNum: sections[indexPath.row].courseId).observe { result in
+                switch result{
```
Comment (danielVebman) [inline CourseGrab/ViewControllers/Home/HomeTableViewController.swift:315]:
```suggestion
                switch result {
```

File: CourseGrab/ViewControllers/Home/HomeTableViewController.swift [hunk 2]
```diff
@@ -295,6 +297,34 @@ extension HomeTableViewController {
...
+                case .value(let response):
+                    guard response.success else{ return }
```
Comment (danielVebman) [inline CourseGrab/ViewControllers/Home/HomeTableViewController.swift:317]:
```suggestion
                    guard response.success else { return }
```

**Review by danielVebman — CHANGES_REQUESTED** [review]
Overall great work! I've got many little nits--watch your white space! If you'd like a quick refresher on our whitespace conventions, leave a comment here and/or ask Haiying. Maybe consider installing SwiftLint too.

File: CourseGrab/ViewControllers/Home/HomeTableViewCell.swift [hunk 3]
```diff
@@ -131,6 +140,17 @@ class HomeTableViewCell: UITableViewCell {
             make.top.leading.equalTo(18)
             make.trailing.equalToSuperview().inset(40)
         }
+        
+        popularityImageView.snp.makeConstraints { make in
+            make.top.equalTo(courseLabel.snp.bottom).offset(9)
+            make.trailing.equalTo(popularityLabel.snp.leading).offset(-6)
```
Comment (Npikielny) [inline CourseGrab/ViewControllers/Home/HomeTableViewCell.swift:146]: Inset?

File: CourseGrab/ViewControllers/Home/HomeTableViewCell.swift [hunk 4]
```diff
@@ -161,6 +183,15 @@ class HomeTableViewCell: UITableViewCell {
             }
         }
     }
+    func changeBackground(isHighlighted: Bool){
+        if isHighlighted{
+            self.containerView.backgroundColor = UIColor(red: 220/255, green: 220/255, blue: 220/255, alpha: 1)
+        }
+        else{
+         self.containerView.backgroundColor = .white
```
Comment (Npikielny) [inline CourseGrab/ViewControllers/Home/HomeTableViewCell.swift:191]: You don't need self and make the else inline. Also think about color literals : )

File: CourseGrab/ViewControllers/Home/HomeTableViewController.swift [hunk 5]
```diff
@@ -295,6 +297,34 @@ extension HomeTableViewController {
...
+    }
+    
+    override func tableView(_ tableView: UITableView, didUnhighlightRowAt indexPath: IndexPath) {
+            let cell = tableView.cellForRow(at: indexPath) as! HomeTableViewCell
+            cell.changeBackground(isHighlighted: false)
+    }
+    
+    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
+        switch tableSections[indexPath.section] {
+        case .available(let sections), .awaiting(let sections):
```
Comment (Npikielny) [inline CourseGrab/ViewControllers/Home/HomeTableViewController.swift:313]: fancy 🤩

**Review by Npikielny — APPROVED** [review]
Looks good to me! Have a happy Thanksgiving : )

File: CourseGrab/Network/Endpoints.swift [hunk 6]
```diff
@@ -84,5 +84,9 @@ extension Endpoint {
         let body = EnableNotificationsBody(notification: enabled ? "IOS" : "NONE")
         return Endpoint(path: "/users/notification/", headers: standardHeaders, body: body)
     }
-
+    
+    static func getCourse(courseNum: Int) -> Endpoint{
```
Comment (danielVebman) [inline CourseGrab/Network/Endpoints.swift:88]: Space before `{`

File: CourseGrab/Network/NetworkManager.swift [hunk 7]
```diff
@@ -59,6 +59,10 @@ class NetworkManager {
         validateToken()
             .chained { self.networking(Endpoint.enableNotifications(enabled: enabled)).decode() }
     }
+    func getCourse(courseNum: Int)-> Future<Response<Course>> {
```
[the remark this answers is not in the document]
Comment (danielVebman) [reply CourseGrab/Network/NetworkManager.swift:63]: Space before `->`

File: CourseGrab/ViewControllers/Home/HomeTableViewCell.swift [hunk 8]
```diff
@@ -161,6 +184,15 @@ class HomeTableViewCell: UITableViewCell {
             }
         }
     }
+    func setBackgroundColor(isHighlighted: Bool){
+        if isHighlighted{
+            self.containerView.backgroundColor = UIColor(red: 220/255, green: 220/255, blue: 220/255, alpha: 1)
+        }
+        else{
```
Comment (danielVebman) [inline CourseGrab/ViewControllers/Home/HomeTableViewCell.swift:191]: Lots of spacing issues here

```suggestion
    func setBackgroundColor(isHighlighted: Bool) {
        if isHighlighted {
            containerView.backgroundColor = UIColor(red: 220 / 255, green: 220 / 255, blue: 220 / 255, alpha: 1)
        } else {
```

And anyway, I suggest this more idiomatic approach:

```
    var isHighlighted: Bool {
        didSet {
            let highlightColor = UIColor(red: 220/255, green: 220/255, blue: 220/255, alpha: 1)
            containerView.backgroundColor = isHighlighted ? highlightColor : .white
        }
```

File: CourseGrab/ViewControllers/Home/HomeTableViewController.swift [hunk 9]
```diff
@@ -295,6 +297,34 @@ extension HomeTableViewController {
...
     }
+    
+    override func tableView(_ tableView: UITableView, didHighlightRowAt indexPath: IndexPath) {
+        let cell = tableView.cellForRow(at: indexPath) as! HomeTableViewCell
+        cell.setBackgroundColor(isHighlighted: true)
+    }
+    
+    override func tableView(_ tableView: UITableView, didUnhighlightRowAt indexPath: IndexPath) {
+            let cell = tableView.cellForRow(at: indexPath) as! HomeTableViewCell
+            cell.setBackgroundColor(isHighlighted: false)
```
Comment (danielVebman) [inline CourseGrab/ViewControllers/Home/HomeTableViewController.swift:308]: You indented too many times here. It should just be 1 level

**Review by danielVebman — CHANGES_REQUESTED** [review]
Any update here, Elvis? Still lots of spacing issues and a couple other points.

File: CourseGrab/Network/Endpoints.swift [hunk 10]
```diff
@@ -84,5 +84,9 @@ extension Endpoint {
         let body = EnableNotificationsBody(notification: enabled ? "IOS" : "NONE")
         return Endpoint(path: "/users/notification/", headers: standardHeaders, body: body)
     }
-
+    
+    static func getCourse(courseNum: Int) -> Endpoint {
+        return Endpoint(path: "/courses/\(courseNum)/", headers: standardHeaders)
```
Comment (danielVebman) [inline CourseGrab/Network/Endpoints.swift:89]: Delete `return`. It's implicit.

File: CourseGrab/ViewControllers/Home/HomeTableViewCell.swift [hunk 11]
```diff
@@ -143,10 +143,11 @@ class HomeTableViewCell: UITableViewCell {
         
         popularityImageView.snp.makeConstraints { make in
             make.top.equalTo(courseLabel.snp.bottom).offset(9)
-            make.trailing.equalTo(popularityLabel.snp.leading).offset(-6)
+            make.trailing.equalTo(popularityLabel.snp.leading).inset(-4)
```
Comment (danielVebman) [inline CourseGrab/ViewControllers/Home/HomeTableViewCell.swift:146]: You went from `offset(-6)` to `inset(-4)`. Is this what you mean to do? Did you not mean `offset(-6)` to `inset(6)` or `inset(4)?

File: CourseGrab/ViewControllers/Home/HomeTableViewCell.swift [hunk 12]
```diff
@@ -183,6 +183,10 @@ class HomeTableViewCell: UITableViewCell {
             }
         }
     }
+    
+    func setBackgroundColor(isHighlighted: Bool){
+        isHighlighted ? (containerView.backgroundColor =  [#colorLiteral(red: 0.8064131141, green: 0.8065492511, blue: 0.8063951135, alpha: 1)].first) : (containerView.backgroundColor = .white)
+    }
```
Comment (danielVebman) [inline CourseGrab/ViewControllers/Home/HomeTableViewCell.swift:189]: I feel like I commented this before but idk where it went. This is better:

```suggestion
    var isHighlighted: Bool {
        didSet {
            containerView.backgroundColor =  isHighlighted ? #colorLiteral(red: 0.8064131141, green: 0.8065492511, blue: 0.8063951135, alpha: 1) : .white
        }
    }
```

Also, do we use color literals anywhere else? If not, I would avoid breaking the convention.

File: CourseGrab/ViewControllers/Home/HomeTableViewController.swift [hunk 13]
```diff
@@ -30,7 +30,7 @@ class HomeTableViewController: UITableViewController {
     private let impactFeedbackGenerator = UIImpactFeedbackGenerator(style: .light)
     private let reachability = try! Reachability()
     private var state: State = .loading
-    private var tableSections: [TableSection] = []
+    private var tableSections: [TableSection] = []
```
Comment (danielVebman) [inline CourseGrab/ViewControllers/Home/HomeTableViewController.swift:33]: Delete the trailing whitespace.

**Review by danielVebman — APPROVED** [review]
Good stuff!

File: CourseGrab/ViewControllers/Home/HomeTableViewCell.swift (hunk 12 above)
Comment (Npikielny) [reply CourseGrab/ViewControllers/Home/HomeTableViewCell.swift:189]: Sadge

File: CourseGrab/ViewControllers/Home/HomeTableViewController.swift [hunk 14]
```diff
@@ -295,6 +295,34 @@ extension HomeTableViewController {
...
+    
+    override func tableView(_ tableView: UITableView, didHighlightRowAt indexPath: IndexPath) {
+        let cell = tableView.cellForRow(at: indexPath) as! HomeTableViewCell
+        cell.setBackgroundColor(isHighlighted: true)
+    }
+    
+    override func tableView(_ tableView: UITableView, didUnhighlightRowAt indexPath: IndexPath) {
+        let cell = tableView.cellForRow(at: indexPath) as! HomeTableViewCell
+        cell.setBackgroundColor(isHighlighted: false)
+    }
```
Comment (Npikielny) [inline CourseGrab/ViewControllers/Home/HomeTableViewController.swift:307]: @danielVebman 
Can we refactor all of this code into the HomeTableViewCell with variable properties for isSelected? 
Something like override var isSelected: Bool {
didSet {
backgroundColor = isSelected ? ... : ...
}
}

File: CourseGrab/Models/Section.swift [hunk 15]
```diff
@@ -56,6 +56,7 @@ struct Section: Codable {
 
     let catalogNum: Int
     let courseNum: Int
+    let courseId: Int
```
Comment (haiyingweng) [inline CourseGrab/Models/Section.swift:59]: `courseId` should go before `courseNum` alphabetically

**Review by haiyingweng — APPROVED** [review]
awesome job 🚀

File: CourseGrab/ViewControllers/Home/HomeTableViewCell.swift (hunk 11 above)
Comment (emarc0314) [reply CourseGrab/ViewControllers/Home/HomeTableViewCell.swift:146]: `inset(4)` actually causes the people image to overlap with the number representing the count of people tracking a course, so instead I use `inset(-4)`.

File: CourseGrab/ViewControllers/Home/HomeTableViewCell.swift [hunk 16]
```diff
@@ -183,6 +183,12 @@ class HomeTableViewCell: UITableViewCell {
             }
         }
     }
+    
+    override var isHighlighted: Bool {
+        didSet{
```
Comment (danielVebman) [inline CourseGrab/ViewControllers/Home/HomeTableViewCell.swift:188]: Space before `{`

File: CourseGrab/ViewControllers/Home/HomeTableViewCell.swift [hunk 17]
```diff
@@ -183,6 +183,12 @@ class HomeTableViewCell: UITableViewCell {
...
+            containerView.backgroundColor =  isHighlighted ? UIColor(red: 220 / 255, green: 220 / 255, blue: 220 / 255, alpha: 1) : .white
```
Comment (danielVebman) [inline CourseGrab/ViewControllers/Home/HomeTableViewCell.swift:189]: You have a double space 

```suggestion
            containerView.backgroundColor = isHighlighted ? UIColor(red: 220 / 255, green: 220 / 255, blue: 220 / 255, alpha: 1) : .white
```

**Review by danielVebman — APPROVED** [review]
Good job! Only some tiny comments left. I strongly recommend you install SwiftLint. It'll catch all these white space and other style issues.

**Pull request merged**